File: python/solutions/day_01.py
FILE_PATH: str = "./../../csharp/InputFiles/01.txt"
TEST_FILE_PATH: str = "./../../csharp/InputFiles/01TestFile.txt"

def day_01(use_test_files: bool = False):
    file_path = get_input_file(use_test_files)
    part_1(file_path)
    part_2(file_path)

def part_1(file_path):
    lhs, rhs = open_input_file(file_path)
    lhs.sort()
    rhs.sort()

    result = sum(abs(lvalue - rvalue) for lvalue, rvalue in zip(lhs, rhs))

    print(f'Day 01 Part 1 - {result}')

def part_2(file_path):
    lhs, rhs = open_input_file(file_path)

    result = sum(value * rhs.count(value) for value in lhs)

    print(f'Day 01 Part 2 - {result}')

def open_input_file(file_path: str) -> ([], []):
    lhs = []
    rhs = []
    with open(file_path, 'r') as file:
        for line in file:
            pairs = line.split()
            lhs.append(int(pairs[0]))
            rhs.append(int(pairs[1]))
    return lhs, rhs

def get_input_file(use_test_files: bool = False):
    if use_test_files:
        return TEST_FILE_PATH
    else:
        return FILE_PATH

File: csharp/Solutions/Day01.cs
using AdventOfCode.InputFiles;
using System.Linq;

namespace AdventOfCode.Solutions
{
    public static class Day01
    {
        private const int Day = 1;

        public static class Part1
        {
            public static double GetAnswer()
            {
                string[] inputData = InputCache.Get(Day, false);
                return Shared.Result(inputData);
            }
        }

        public static class Part2
        {
            public static double GetAnswer()
            {
                string[] inputData = InputCache.Get(Day, false);
                return Shared.Result(inputData, true);
            }
        }

        private class Shared
        {
            public static double Result(string[] set, bool partTwo = false)
            {
                var result = 0d;

                var lhs = new List<int>();
                var rhs = new List<int>();

                for (int i = 0; i < set.Length; i++)
                {
                    if (set[i].Trim() == string.Empty)
                    {
                        continue;
                    }
                    var pairs = set[i].Split("   ");
                    lhs.Add(Convert.ToInt32(pairs[0]));
                    rhs.Add(Convert.ToInt32(pairs[1]));
                }

                if (!partTwo)
                {
                    lhs.Sort();
                    rhs.Sort();

                    foreach (var (lValue, rValue) in lhs.Zip(rhs, (a, b) => (a, b)))
                    {
                        result += Math.Abs(lValue - rValue);
                    }
                }
                else
                {
                    foreach(var value in lhs)
                    {
                        result += value * rhs.Count(v => v == value);
                    }
                }

                return result;
            }
        }
    }
}
